### DISCUSSION TITLE
Recognize auth changes in Svelte

### DISCUSSION THREAD

**Discussion by jimafisk** [pocketbase/pocketbase#217] [Q&A] [answered]
I'm trying to get a Svelte SPA to update components when the PocketBase `client` updates. For example, I'm checking `client.AuthStore.isValid` and if it's `false` I'll show a "Sign In" button and it it's `true` I'll show a "Logout" button. PocketBase is correctly updating the `client` value behind the scenes when you Sign In / Logout, you can see it with a simple interval:
```js
onMount(async () => {
  client = new PocketBase('http://localhost:8090');
  setInterval(() => {
    console.log(client);
  }, 5000);
});
```
However, I have to do a hard refresh manually to toggle the Sign In / Logout button in the UI because Svelte is not recognizing an update to the `client` prop. Supabase has [onAuthStateChange()](https://supabase.com/docs/reference/javascript/auth-onauthstatechange) for this, here's an example of it being used in a Svelte app: https://supabase.com/docs/guides/with-svelte#launch. Is there a way to accomplish something similar in PocketBase? Thank you!

**Comment by ganigeorgiev** [comment, marked as answer]
Similar thing was asked in https://github.com/pocketbase/pocketbase/issues/195, but initially it wasn't clear to me what actually the user wanted to do.

Now I think I understand the request better and it shouldn't be very difficult to add some kind of listeners support to the `AuthStore`.
While working on the Dart SDK, I also noticed some other minor things that we can improve in the JavaScript implementation so I'll rework it alongside the other changes I'm planing somewhere around next week.

_As a workaround for now, you could do something similar like in the Admin UI. Basically I have an `admin` Svelte store that I'm updating in a slightly extended `client.AuthStore` implementation - https://github.com/pocketbase/pocketbase/blob/master/ui/src/utils/ApiClient.js#L67_

**Reply by jimafisk** [reply]
That worked! Having a listener would probably make this easier, thanks for considering it for you roadmap.

For folks who need a little help, the workaround could look something like this: https://svelte.dev/repl/73e9be09054c4b01aeb5baf5d5bdfe83?version=3.49.0

**Reply by ganigeorgiev** [reply]
@jimafisk Just to let you know that support for `onChange(fn)` listener was added in [v0.3.0 release of the Javascript SDK](https://github.com/pocketbase/js-sdk/releases/tag/v0.3.0):
```js
client.authStore.onChange((token, model) => {
    console.log("Store changed":, token, model);
})
```

The AuthStore documentation was also updated and a new `BaseAuthStore` was added to make it easier for users to create their own custom store instances - https://github.com/pocketbase/js-sdk#authstore.

**Reply by jimafisk** [reply]
This is amazing @ganigeorgiev, it works great! Thank you!

**Comment by MedLabs** [comment]
I'm using js-sdk v 0.3.0 and onChange still does not exist in authStore :/

**Reply by ganigeorgiev** [reply]
@MedLabs Are you using the default [`LocalAuthStore`](https://github.com/pocketbase/js-sdk/blob/master/src/stores/LocalAuthStore.ts) or your own custom store?

If it is the latter, make sure that your store extends [`BaseAuthStore`](https://github.com/pocketbase/js-sdk/blob/master/src/stores/BaseAuthStore.ts) and calls its respective `super.save()` and `super.clear()` methods.
